src/supervisor-api.ts

```diff
@@ -0,0 +1,196 @@
+			const conf = await config.getMany([
+				'apiSecret',
+				'localMode',
+				'unmanaged',
+				'osVariant',
+			]);
+
+			const needsAuth = conf.unmanaged
+				? conf.osVariant === 'prod'
+				: !conf.localMode;
+
+			if (needsAuth) {
+				// Only get the key if we need it
+				const key = getKeyFromReq(req);
+				if (key && conf.apiSecret && key === conf.apiSecret) {
+					return next();
+				} else {
+					return res.sendStatus(401);
+				}
+			} else {
+				return next();
+			}
+		} catch (err) {
+			res.status(503).send(`Unexpected error: ${err}`);
+		}
+	};
+}
+
+const expressLogger = morgan((tokens, req, res) =>
+	[
+		'Supervisor API:',
+		tokens.method(req, res),
+		req.path,
+		tokens.status(req, res),
+		'-',
+		tokens['response-time'](req, res),
+		'ms',
+	].join(' '),
+);
+
+interface SupervisorAPIConstructOpts {
+	config: Config;
+	eventTracker: EventTracker;
+	routers: express.Router[];
+	healthchecks: Array<() => Promise<boolean>>;
+}
+
+export class SupervisorAPI {
+	private config: Config;
+	private eventTracker: EventTracker;
+	private routers: express.Router[];
+	private healthchecks: Array<() => Promise<boolean>>;
+
+	private api = express();
+	private server: Server | null = null;
+
+	public constructor({
+		config,
+		eventTracker,
+		routers,
+		healthchecks,
+	}: SupervisorAPIConstructOpts) {
+		this.config = config;
+		this.eventTracker = eventTracker;
+		this.routers = routers;
+		this.healthchecks = healthchecks;
+
+		this.api.disable('x-powered-by');
+		this.api.use(expressLogger);
+		this.api.use(authenticate(this.config));
+
+		this.api.get('/v1/healthy', async (_req, res) => {
+			try {
+				const healths = await Promise.all(this.healthchecks.map(fn => fn()));
+				if (!_.every(healths)) {
+					throw new Error('Unhealthy');
+				}
+				return res.sendStatus(200);
+			} catch (e) {
+				res.sendStatus(500);
+			}
+		});
+
+		this.api.get('/ping', (_req, res) => res.send('OK'));
+
+		this.api.post('/v1/blink', (_req, res) => {
+			this.eventTracker.track('Device blink');
+			blink.pattern.start();
+			setTimeout(blink.pattern.stop, 15000);
+			return res.sendStatus(200);
+		});
+
+		// Expires the supervisor's API key and generates a new one.
+		// It also communicates the new key to the balena API.
+		this.api.post('/v1/regenerate-api-key', async (_req, res) => {
+			try {
+				const secret = await this.config.newUniqueKey();
+				await this.config.set({ apiSecret: secret });
+				res.status(200).send(secret);
+			} catch (e) {
+				res
+					.status(503)
+					.send(e != null ? e.message : undefined || e || 'Unknown error');
+			}
+		});
+
+		// And assign all external routers
+		for (const router of this.routers) {
+			this.api.use(router);
+		}
+	}
+
+	public async listen(
+		allowedInterfaces: string[],
+		port: number,
+		apiTimeout: number,
+	): Promise<void> {
+		const localMode = (await this.config.get('localMode')) || false;
+		await this.applyListeningRules(
+			checkTruthy(localMode) || false,
```

Comment:
For the same reasons as https://github.com/balena-io/balena-supervisor/pull/860#discussion_r244699290 the type system thinks it can. This will all be fixed as soon as I cover all the uses with the conversion.